### PR TITLE
SAPI-437-Frontend service for kafka resilience

### PR DESCRIPTION
**Pull request by bpathak-ch**
* stream request handler
* healthcheck handler
* eric authentication
* TODO implementation with cache broker

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,3 +2,14 @@ module github.com/companieshouse/chs-streaming-api-frontend
 
 go 1.12
 
+require (
+	github.com/Shopify/sarama v1.27.0
+	github.com/companieshouse/chs-go-avro-schemas v0.1.0
+	github.com/companieshouse/chs-streaming-api v2.3.1-rc1+incompatible
```
Comment (bpathak-ch) [inline go.mod]: any reference to chs-streaming-api has to be removed

**Comment by dlloyd1-ch** [comment]
Quite a lot of changes for one PR. Could it be broken down into components defined in the LLD (Streaming request handler, authentication handler etc)?

File: config/config.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/companieshouse/gofigure"
+)
+
+// Config represents the frontend application configuration.
+type Config struct {
+	gofigure            interface{} `order:"env,flag"`
+	BindAddr            string      `env:"BIND_ADDR"                    flag:"bind-addr"              flagDesc:"Bind address"`
+	CertFile            string      `env:"CERT_FILE"                    flag:"cert-file"              flagDesc:"Certificate file"`
+	EricURL             string      `env:"ERIC_LOCAL_URL"               flag:"eric-url"               flagDesc:"Eric url"`
+	FilingTopic         string      `env:"FILING_TOPIC"                 flag:"filing-topic"           flagDesc:"Filing stream topic"`
```
Comment (dlgroves-ch) [inline config/config.go]: Not required - Kafka integration should be handled by chs-streaming-api-backend.

File: config/config.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	HeartbeatInterval   int         `env:"HEARTBEAT_INTERVAL"           flag:"heartbeat-interval"     flagDesc:"Heartbeat interval in seconds"`
+	KeyFile             string      `env:"KEY_FILE"                     flag:"key-file"               flagDesc:"Key file"`
+	LogTopic            string      `env:"LOG_TOPIC"                    flag:"kafka-log-topic"        flagDesc:"Log topic"`
```
Comment (dlgroves-ch) [inline config/config.go]: Not used

File: config/config.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	RequestTimeout      int         `env:"REQUEST_TIMEOUT"              flag:"request-timeout"        flagDesc:"Request timeout in seconds"`
+	CacheBrokerURL      string      `env:"CACHE_BROKER_URL"          	flag:"cache-broker-url"    	  flagDesc:"Cache broker url"`
+	StreamingBrokerAddr []string    `env:"KAFKA_STREAMING_BROKER_ADDR"  flag:"streaming-broker-addr"  flagDesc:"Streaming CH Kafka broker cluster address"`
```
Comment (dlgroves-ch) [inline config/config.go]: Not required - Kafka integration should be handled by chs-streaming-api-backend.

File: consumer/consumer.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
+// Package consumer provides an initialiser for a message consumer.
+package consumer
```
Comment (dlgroves-ch) [inline consumer/consumer.go]: This whole package can go as chs-streaming-api-frontend won't be connected directly to Kafka.

File: handlers/stream.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+var callLogHandleTimeOut = handleTimeOut
+var callLogHandleClientDisconnect = handleClientDisconnect
+var callProcessMessage = ProcessMessage
+var callProcessMessageFailed = processMessageFailed
+var callLogHeartbeatTimeout = logHeartbeatTimeout
+var callCheckIfDataNotPresent = checkIfDataNotPresent
+var callResponseWriterWrite = responseWriterWrite
+
+// ProcessHTTP Test at this level for unit testing.
+func (st Streaming) ProcessHTTP(topic string, w http.ResponseWriter, req *http.Request, stream Stream) {
```
Comment (dlgroves-ch) [inline handlers/stream.go]: Given that the frontend will now be decoupled from Kafka, this handler should do two things:

1. Fetch a range of offsets cached by chs-streaming-api-cache if a timepoint has been specified.
2. Stream offsets published to it by chs-streaming-api-cache to the connected user.

File: offset/offset.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
+// Package offset provides functions for validating request offsets.
+package offset
```
Comment (dlgroves-ch) [inline offset/offset.go]: Not required - chs-streaming-api-cache should be responsible for reporting whether or not a given offset number is valid and the frontend should be responsible for handling the error that it returns (if any).

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

File: handlers/stream.go (hunk 6 above)
Comment (bpathak-ch) [reply handlers/stream.go]: noted and added as comment for the next PR

File: routes.yaml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+app_name: chs-streaming-api-frontend
```
Comment (dlgroves-ch) [inline routes.yaml]: Not required; we'll be deploying this application using ECS.

File: handlers/filing/filing.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package filing
```
Comment (dlgroves-ch) [inline handlers/filing/filing.go]: Not required

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

File: routes.yaml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+group: streampath
```
Comment (dlgroves-ch) [inline routes.yaml]: Please remove this file; we will be deploying this application to ECS as a container.

**Review by dlgroves-ch — CHANGES_REQUESTED** [review]

**Review by dlgroves-ch — APPROVED** [review]

**Pull request merged**